internal/retry/retry.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package retry provides retry logic.
+package retry
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	gax "github.com/googleapis/gax-go"
+)
+
+// Call calls the supplied function f repeatedly, using the isRetryable function and
+// the provided backoff parameters to control the repetition.
+//
+// When f returns nil, Call immediately returns nil.
+//
+// When f returns an error for which isRetryable returns false, Call immediately
+// returns that error.
+//
+// When f returns an error for which isRetryable returns true, Call sleeps for the
+// duration provided by isRetryable, or the provided backoff value if that duration
+// is zero, and invokes f again.
+//
+// When the provided context is done, Retry returns with an error that
```

Comment:
I would suggest "an error" -> "a ContextError" to make it more clear that it's a special error type; when I first read this I was confused by what it meant for an error to include 2 other errors.

Reply:
Done in new PR.